In [11]:
import pandas as pd

X_train = pd.read_csv('./datasets/mod_04_hw_train_data.csv')
X_test = pd.read_csv('./datasets/mod_04_hw_valid_data.csv')

In [12]:
y_train = X_train['Salary']
y_test = X_test['Salary']

In [13]:
X_train = X_train.drop(['Name', 'Phone_Number', 'Salary', 'Date_Of_Birth', 'Qualification'], axis=1)
X_test = X_test.drop(['Name', 'Phone_Number', 'Salary', 'Date_Of_Birth', 'Qualification'], axis=1)

In [14]:
X_train_cat = X_train.select_dtypes(include='object')
X_train_num = X_train.select_dtypes(exclude='object')

In [15]:
X_test_cat = X_test.select_dtypes(include='object')
X_test_num = X_test.select_dtypes(exclude='object')

In [16]:
from sklearn.impute import SimpleImputer

imputer = SimpleImputer(strategy='most_frequent').set_output(transform='pandas')

X_train_num = imputer.fit_transform(X_train_num)
X_test_num = imputer.transform(X_test_num)

In [17]:
from sklearn.preprocessing import PowerTransformer 

transformer = PowerTransformer().set_output(transform='pandas')

X_train_num = transformer.fit_transform(X_train_num)
X_test_num = transformer.transform(X_test_num)

In [18]:
from sklearn.preprocessing import OneHotEncoder 

encoder = (OneHotEncoder(drop='if_binary',
                         sparse_output=False,)
           .set_output(transform='pandas'))

X_train_cat = encoder.fit_transform(X_train_cat)
X_test_cat = encoder.transform(X_test_cat)

In [19]:
from sklearn.neighbors import KNeighborsRegressor

regressor = KNeighborsRegressor(n_neighbors=5)
regressor.fit(pd.concat([X_train_cat, X_train_num], axis=1), y_train)

y_pred = regressor.predict(pd.concat([X_test_cat, X_test_num], axis=1))

In [20]:
from sklearn.metrics import mean_absolute_percentage_error

mape = mean_absolute_percentage_error(y_test, y_pred)

print(f'Validation MAPE: {mape:.2%}')

Validation MAPE: 4.68%


### Висновок
Результат моєї роботи відповідає очікуваному діапазону MAPE (3-5%), отриманий MAPE становить 4.68%. Я виконала усі необхідні кроки для побудови моделі, включаючи обробку даних та налаштування моделі.

### Опис кроків:

1.  Імпорт даних

2. Витягнула цільову змінну 'Salary' з обох наборів даних.

3. Видалила стовпці, які не потрібні для моделювання: 'Name', 'Phone_Number', 'Salary', 'Date_Of_Birth', 'Qualification'

4. Поділ на числові та категоріальні змінні:
Розділила числові та категоріальні ознаки в окремі DataFrame.

5. Заповнення пропущених значень:
Заповнила пропущені значення в числових ознаках значеннями моди.

6. Нормалізація даних:
Виконала нормалізацію числових ознак за допомогою PowerTransformer.

7. Кодування категоріальних змінних:
Закодувала категоріальні ознаки методом OneHotEncoder.

8. Створила та навчила модель KNeighborsRegressor з 5 сусідами.

9. Прогнозувала заробітну плату для тестового набору та обчислила MAPE для оцінки моделі.

Validation MAPE: 4.68%
